### PR TITLE
Add sorting script previews by modified date

### LINKED ISSUE #8602: Sort script list by the scripts' modification date
### Describe the project you are working on

Any game or plugin with some amount of scripting involved.

### Describe the problem or limitation you are having in your project

I often end up with a lot of open scripts populating the side scripts preview window next to the code editor. I don't want to close them all the time, but as the number of items in the small window grows, it's hard to stay organized. When working with multiple files simultaneously (like developing a class hierarchy), it can be hard to navigate to the correct files in the preview.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

I'd like to add a new type of sorting to the script previews - sort by modified date. This would keep the most recently modified files at the top of the preview list.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

When a script is saved, it jumps to the top of the preview list.

https://github.com/godotengine/godot-proposals/assets/49688205/72455a67-c13c-410a-a260-8a9675bac9d5

I actually put the horse before the cart by mistake here and implemented the feature in [this pull request](https://github.com/godotengine/godot/pull/85861). That's also where the video is from.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

This cannot be done with an external script.

### Is there a reason why this should be core and not an add-on in the asset library?

I don't think this is possible with an add-on.

### PR DESCRIPTION
**Pull request by VojtaStruhar**
Adds an option for sorting the scripts preview by modified time (descending), so that the latest modified files appear on top.

The default sorting by name works well, but often I have a lot of scripts open and I want to quickly jump between the last few I'm working on without having to remember which one is which. This pull request adds this functionality. In the video I modify one of the scripts, press `Cmd + S` and it jumps to the top of the preview.

https://github.com/godotengine/godot/assets/49688205/7474c0c7-1644-4b10-bfd2-b9e1953f371c

## Consideration: The scripts are sorted on save

As you can see in the code changes, the _actual sorting_ of the script items inside the preview now happens every time the script is saved (`ScriptEditor::_res_saved_callback`). For the other sorting methods this is absolutely **not** needed, but is performed anyway. I think this is not a performance bottleneck (there will likely not be _thousands_ of scripts in the preview at any one time and they are saved/sorted on demand with a user action). Since this is an editor thing, it doesn't affect games at all.

One solution to this would be to check if we are in the _SORT_BY_MODIFIED_ mode and only perform the sorting then, but that seems dirty to me. Stray if conditions like that are not very maintainable imo.

---

**This is my first real contribution**, so please double check if the changes are sane. If naming is OK, if I'm using things for what they are supposed to be used, etc.

- *Production edit: This closes https://github.com/godotengine/godot-proposals/issues/8602.*

### CODE REVIEW

**Comment by Calinou** [comment]
Thanks for opening a pull request :slightly_smiling_face: 

Feature pull requests should be associated to a *feature proposal* to justify the need for implementing the feature in Godot core. Please open a proposal on the [Godot proposals](https://github.com/godotengine/godot-proposals) repository (and link to this pull request in the proposal's body).

**Comment by VojtaStruhar** [comment]
Sorry about that! Will do

File: editor/plugins/script_editor_plugin.cpp [hunk 1]
```diff
@@ -2063,6 +2064,11 @@ void ScriptEditor::_update_script_names() {
 				case SORT_BY_PATH: {
 					sd.sort_key = path;
 				} break;
+				case SORT_BY_MODIFIED: {
+					// XOR with UINT64_MAX results in descending order
```
Comment (AThousandShips) [inline editor/plugins/script_editor_plugin.cpp:2068]:
```suggestion
					// XOR with UINT64_MAX results in descending order.
```
Per the [comment style](https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html#comment-style-guide).

File: editor/plugins/script_editor_plugin.h [hunk 2]
```diff
@@ -255,6 +255,7 @@ class ScriptEditor : public PanelContainer {
 	enum ScriptSortBy {
 		SORT_BY_NAME,
 		SORT_BY_PATH,
+		SORT_BY_MODIFIED,
```
Comment (AThousandShips) [inline editor/plugins/script_editor_plugin.h:258]:
```suggestion
		SORT_BY_MODIFIED_TIME,
```
To match `FilesystemDock`

File: editor/plugins/script_editor_plugin.cpp [hunk 3]
```diff
@@ -4373,7 +4379,7 @@ ScriptEditorPlugin::ScriptEditorPlugin() {
 	EDITOR_DEF("text_editor/script_list/script_temperature_enabled", true);
 	EDITOR_DEF("text_editor/script_list/script_temperature_history_size", 15);
 	EDITOR_DEF("text_editor/script_list/group_help_pages", true);
-	EditorSettings::get_singleton()->add_property_hint(PropertyInfo(Variant::INT, "text_editor/script_list/sort_scripts_by", PROPERTY_HINT_ENUM, "Name,Path,None"));
+	EditorSettings::get_singleton()->add_property_hint(PropertyInfo(Variant::INT, "text_editor/script_list/sort_scripts_by", PROPERTY_HINT_ENUM, "Name,Path,Modified,None"));
```
Comment (AThousandShips) [inline editor/plugins/script_editor_plugin.cpp:4382]:
```suggestion
	EditorSettings::get_singleton()->add_property_hint(PropertyInfo(Variant::INT, "text_editor/script_list/sort_scripts_by", PROPERTY_HINT_ENUM, "Name,Path,Last Modified Time,None"));
```
Same, and more clear

**Comment by VojtaStruhar** [comment]
@AThousandShips thanks for the feedback!!!

File: editor/plugins/script_editor_plugin.cpp [hunk 4]
```diff
@@ -2063,6 +2064,11 @@ void ScriptEditor::_update_script_names() {
 				case SORT_BY_PATH: {
 					sd.sort_key = path;
 				} break;
+				case SORT_BY_MODIFIED: {
```
Comment (AThousandShips) [inline editor/plugins/script_editor_plugin.cpp:2067]:
```suggestion
				case SORT_BY_MODIFIED_TIME: {
```
You forgot to update this one

Comment (VojtaStruhar) [reply editor/plugins/script_editor_plugin.cpp]: Ooops, missed one. Thanks.

**Comment by VojtaStruhar** [comment]
I just realized, that the fact it's sorting on _every script save_ will interfere with the `Alt + Shift + Up/Down` keybind that moves the current script up or down in the script preview window. This is definitely not what we want for other sorting styles. Any suggestions for an elegant solution to this?
- Adding another settings checkbox to _sort on script update_ is overkill
- Stray standalone if check for this sorting mode doesn't feel right to me :/

**Pull request closed without merging**